docs/architecture/adr-013-msg-authorization.md

```diff
@@ -0,0 +1,220 @@
+// account with the provided expiration time.
+type MsgGrant struct {
+	Granter    sdk.AccAddress `json:"granter"`
+	Grantee    sdk.AccAddress `json:"grantee"`
+	Capability Capability     `json:"capability"`
+    // Expiration specifies the expiration time of the grant
+	Expiration time.Time      `json:"expiration"`
+}
+```
+
+#### MsgRevoke
+
+```go
+// MsgRevoke revokes any capability with the provided sdk.Msg type on the
+// granter's account with that has been granted to the grantee.
+type MsgRevoke struct {
+	Granter sdk.AccAddress `json:"granter"`
+	Grantee sdk.AccAddress `json:"grantee"`
+    // CapabilityMsgType is the type of sdk.Msg that the revoked Capability refers to.
+    // i.e. this is what `Capability.MsgType()` returns
+	CapabilityMsgType sdk.Msg `json:"capability_msg_type"`
+}
+```
+
+#### MsgExecDelegated
+
+```go
+// MsgExecDelegated attempts to execute the provided messages using
+// capabilities granted to the grantee. Each message should have only
+// one signer corresponding to the granter of the capability.
+type MsgExecDelegated struct {
+	Grantee sdk.AccAddress `json:"grantee"`
+	Msgs   []sdk.Msg      `json:"msg"`
+}
+```
+
+### Keeper
+
+#### Constructor: `NewKeeper(storeKey sdk.StoreKey, cdc *codec.Codec, router sdk.Router) Keeper`
+
+The message delegation keeper receives a reference to the baseapp's `Router` in order
+to dispatch delegated messages.
+
+#### `DispatchActions(ctx sdk.Context, grantee sdk.AccAddress, msgs []sdk.Msg) sdk.Result`
+
+`DispatchActions` attempts to execute the provided messages via capability
+grants from the message signer to the grantee.
+
+#### `Grant(ctx sdk.Context, grantee sdk.AccAddress, granter sdk.AccAddress, capability Capability, expiration time.Time)`
+
+Grants the provided capability to the grantee on the granter's account with the provided expiration
+time. If there is an existing capability grant for the same `sdk.Msg` type, this grant
+overwrites that.
+
+#### `Revoke(ctx sdk.Context, grantee sdk.AccAddress, granter sdk.AccAddress, msgType sdk.Msg)`
+
+Revokes any capability for the provided message type granted to the grantee by the granter.
+
+#### `GetCapability(ctx sdk.Context, grantee sdk.AccAddress, granter sdk.AccAddress, msgType sdk.Msg) (cap Capability, expiration time.Time)`
+
+Returns any `Capability` (or `nil`), with the expiration time, granted to the grantee by the granter for the provided msg type.
+
+### CLI
+
+#### `--send-as` Flag
+
+When a CLI user wants to run a transaction as another user using `MsgExecDelegated`, they
+can use the `--send-as` flag. For instance `gaiacli tx gov vote 1 yes --from mykey --send-as cosmos3thsdgh983egh823`
+would send a transaction like this:
+
+```go
+MsgExecDelegated {
+  Grantee: mykey,
+  Msgs: []sdk.Msg{
+    MsgVote {
+	  ProposalID: 1,
+	  Voter: cosmos3thsdgh983egh823
+	  Option: Yes
+    }
+  }
+}
+```
+#### `tx grant <grantee> <capability> --from <granter>`
```

Comment:
@aaronc @ethanfrey

How about expecting all the required fields as parameters instead. So for SendCapability the cli cmd looks something like:
`tx grant <grantee> send <SpendLimit> --from <granter> --expiration <time>`

generic cli cmd for grant is,
`tx grant <grantee> <msg_type> <...fileds> --from <granter> --expiration <time>`

I think this would work best if we have limited parameters required (I suppose there sould be limited inputs required from user).

Reply:
I am happy with `tx grant <grantee> <msg_type>` as positional parameters.

`--from <granter` is needed by the sdk to sign properly I believe? Otherwise I would make that positional as well.

I think the other msg-specific fields (like send-limit), as well as expiration, should be names arguments, just to make the meaning clearer.

What do you think of that?